src/main/java/net/discordjug/javabot/listener/filter/MessageFilterHandler.java

```diff
@@ -0,0 +1,112 @@
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@RequiredArgsConstructor
+public class MessageFilterHandler extends ListenerAdapter {
+
+    private final List<MessageFilter> filters;
+    private final AutoMod autoMod;
+    private final BotConfig botConfig;
+
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
+        if (!shouldRunFilters(event)) {
+            return;
+        }
+        Message message = event.getMessage();
+        StringBuilder builder = new StringBuilder(message.getContentRaw());
+        List<Message.Attachment> attachments = new ArrayList<>(message.getAttachments());
+        List<MessageEmbed> embeds = new ArrayList<>();
+
+        boolean handled = false;
+
+        MessageContent content = new MessageContent(event, builder, attachments, embeds);
+
+        for (MessageFilter filter : filters) {
+            MessageModificationStatus status = filter.processMessage(content);
+            switch (status) {
+                case MODIFIED -> handled = true;
+                case STOP_PROCESSING -> {
+                    return;
+                }
+            }
+        }
+
+        if (handled) {
+            IWebhookContainer webhookContainer = null;
+            long threadId = 0;
+            if (event.isFromType(ChannelType.TEXT)) {
+                webhookContainer = event.getChannel().asTextChannel();
+            }
+            if (event.isFromThread()) {
+                StandardGuildChannel parentChannel = event.getChannel()
+                        .asThreadChannel()
+                        .getParentChannel()
+                        .asStandardGuildChannel();
+                threadId = event.getChannel().getIdLong();
+                webhookContainer = (IWebhookContainer) parentChannel;
+            }
+            if (webhookContainer == null) {
+                return;
+            }
+            replaceMessage(event, webhookContainer, threadId, content);
+        }
+    }
+
+    private boolean shouldRunFilters(@NotNull MessageReceivedEvent event) {
+        if (event.isWebhookMessage()) {
+            return false;
+        }
+        if (!event.isFromGuild()) {
+            return false;
+        }
+        if (event.getAuthor().isBot() || event.getAuthor().isSystem()) {
+            return false;
+        }
+        if (autoMod.hasSuspiciousLink(event.getMessage()) ||
+                autoMod.hasAdvertisingLink(event.getMessage())) {
+            return false;
+        }
+        if (event.getChannel().getIdLong() == botConfig.get(event.getGuild())
+                .getModerationConfig()
+                .getSuggestionChannelId()) {
+            return false;
+        }
+        return true;
+    }
+
+    private void replaceMessage(@NotNull MessageReceivedEvent event, IWebhookContainer webhookContainer,
```

Comment:
The `event` parameter is redundant as it is also included in the `MessageContent` record.